packages/xsnap/src/object-inspect.js

```diff
@@ -0,0 +1,489 @@
+    return false;
+  }
+  try {
+    setSize.call(x);
+    try {
+      mapSize.call(x);
+    } catch (m) {
+      return true;
+    }
+    return x instanceof Set; // core-js workaround, pre-v2.5.0
+  } catch (e) {}
+  return false;
+}
+
+function isWeakSet(x) {
+  if (!weakSetHas || !x || typeof x !== 'object') {
+    return false;
+  }
+  try {
+    weakSetHas.call(x, weakSetHas);
+    try {
+      weakMapHas.call(x, weakMapHas);
+    } catch (s) {
+      return true;
+    }
+    return x instanceof WeakSet; // core-js workaround, pre-v2.5.0
+  } catch (e) {}
+  return false;
+}
+
+function inspectString(str, opts) {
+  if (str.length > opts.maxStringLength) {
+    const remaining = str.length - opts.maxStringLength;
+    const trailer = `... ${remaining} more character${
+      remaining > 1 ? 's' : ''
+    }`;
+    return (
+      inspectString($slice.call(str, 0, opts.maxStringLength), opts) + trailer
+    );
+  }
+  // eslint-disable-next-line no-control-regex
+  const s = $replace.call(
+    $replace.call(str, /(['\\])/g, '\\$1'),
+    // eslint-disable-next-line no-control-regex
+    /[\x00-\x1f]/g,
+    lowbyte,
+  );
+  return wrapQuotes(s, 'single', opts);
+}
+
+function lowbyte(c) {
+  const n = c.charCodeAt(0);
+  const x = {
+    8: 'b',
+    9: 't',
+    10: 'n',
+    12: 'f',
+    13: 'r',
+  }[n];
+  if (x) {
+    return `\\${x}`;
+  }
+  return `\\x${n < 0x10 ? '0' : ''}${$toUpperCase.call(n.toString(16))}`;
+}
+
+function markBoxed(str) {
+  return `Object(${str})`;
+}
+
+function weakCollectionOf(type) {
+  return `${type} { ? }`;
+}
+
+function collectionOf(type, size, entries, indent) {
+  const joinedEntries = indent
+    ? indentedJoin(entries, indent)
+    : $join.call(entries, ', ');
+  return `${type} (${size}) {${joinedEntries}}`;
+}
+
+function singleLineValues(xs) {
+  for (let i = 0; i < xs.length; i += 1) {
+    if (indexOf(xs[i], '\n') >= 0) {
+      return false;
+    }
+  }
+  return true;
+}
+
+function getIndent(opts, depth) {
+  let baseIndent;
+  if (opts.indent === '\t') {
+    baseIndent = '\t';
+  } else if (typeof opts.indent === 'number' && opts.indent > 0) {
+    baseIndent = $join.call(Array(opts.indent + 1), ' ');
+  } else {
+    return null;
+  }
+  return {
+    base: baseIndent,
+    prev: $join.call(Array(depth + 1), baseIndent),
+  };
+}
+
+function indentedJoin(xs, indent) {
+  if (xs.length === 0) {
+    return '';
+  }
+  const lineJoiner = `\n${indent.prev}${indent.base}`;
+  return `${lineJoiner + $join.call(xs, `,${lineJoiner}`)}\n${indent.prev}`;
+}
+
+function arrObjKeys(obj, inspect) {
+  const isArr = isArray(obj);
+  const xs = [];
+  if (isArr) {
+    xs.length = obj.length;
+    for (let i = 0; i < obj.length; i += 1) {
+      xs[i] = has(obj, i) ? inspect(obj[i], obj) : '';
+    }
+  }
+  const syms = getOwnPropertySymbols(obj);
+  for (const key in obj) {
+    if (!has(obj, key)) {
+      // eslint-disable-next-line no-continue
+      continue;
+    }
```

Comment:
The `for/in` has another problem: it does not have a deterministic standard semantics, in the case where the object and/or its prototype chain are mutated during the iteration. This case is known to differ between engines, and is so highly optimized on each that it has been impossible to get the engines to agree on any deterministic spec.

That's not a killer for us. We live with several of these static spec indeterminacies. In the short term, by requiring all validators to run the same engine. In the long term, by writing down a "Deterministic JS" spec that's a refinement[1] of the JS spec. Or better, by skipping that and going directly to writing a Deterministic Hardened JS spec that's a refinement of the hardened JS spec. (We still must keep the distinction. Hardened JS will continue to run on Node. Deterministic Hardened JS never will.)

[1] "refinement" meaning that anything that conforms to the refined spec also conforms to the original spec, but not vice versa. Everything the refined spec requires, the original spec allows.

Nevertheless, all that said, it is worth some trouble to get rid of all `for/in`s in our sources. This one enumerates inherited and own enumerable string-based properties, and then filters that down to own properties.

Note that `Object.getOwnPropertyNames` is not a semantics preserving replacement, in that it will include non-enumerable string-named own properties as well.

Reply:
I used `getOwnPropertyNames` and `isEnumerable` to check.